orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -122,16 +124,24 @@ def _forward(self, *inputs, **kwargs):
                 self._get_inference_graph_and_init_gradient_graph_builder(
                     *inputs, **kwargs)
 
+            # Flag to indicate whether the gradient_graph needs to be built
+            build_gradient_graph = self._current_input_shape is None
             _, _, input_names_require_grad, new_input_shape = \
                 _ortmodule_io.parse_inputs_for_onnx_export(
                     self._original_module_parameters, self._onnx_inference, *inputs, **kwargs)
+            initializer_names_to_train_set = {initializer[0] for initializer in
+                self._flattened_output_module.named_parameters() if initializer[1].requires_grad}
             # If inputs requiring gradient change from forward to the next, the module_gradient_graph_builder
             # needs to be reinitialized so it can compute the backward output for the new inputs that require_grad
-            if input_names_require_grad != self._input_names_require_grad:
+            if input_names_require_grad != self._input_names_require_grad or \
+                initializer_names_to_train_set != self._initializer_names_to_train_set:
                 self._input_names_require_grad = input_names_require_grad
+                self._initializer_names_to_train_set = initializer_names_to_train_set
```

Comment:
Any way we can avoid having `self._initializer_names_to_train` and `self._initializer_names_to_train_set`?

Reply:
```self._initializer_names_to_train``` is not a direct property. It is one that is accessed through the onnx graph populated by the backend through ```self._onnx_graphs_info.initializer_names_to_train```.
In order to keep the lookup time linear, I need to have a set to look up from. But the order is also important for building the backward results which is why, we need both.